src/os2datascanner/utils/optimiser/optimiser.py

```diff
@@ -0,0 +1,314 @@
+                        try:
+                            cont.parent.components.remove(cont)
+                            self.containers.remove(cont)
+                            cont.parent.components.extend(cont.components)
+                            continue
+                        except ValueError:
+                            pass
+                    if self.check_empty(cont) or self.check_useless(cont) or self.check_symbol_redundancy(cont):
+                        continue
+
+        dump_json(main.as_dict(), self.OUT_PATH)
+        return self.cycles > 1 # If it did more than 1 cycle, it means it found something
+
+    def reformat_rule(self, path):
+        obj = str(load_json(path))
+        for id in rule_ids:
+            obj = obj.replace(str(id), rule_ids[id])
+        obj = obj.replace("'", '"') # Fixing some double | single quote problems
+        obj = json.loads(obj)
+        with open(OUT_PATH, "wt") as file:
+            json.dump(obj, file)
+
+    def run_optimiser(self):
+        refactored = True
+        while refactored:
+            refactored = self.optimisation_cycle()
+            # Get previous cycles. 
+            # While previous optimisation_cycle() found stuff, run it again.
+            # Maybe check for risk of infinite recursion
+        self.reformat_rule(OUT_PATH)
+
+
+class CustomContainer():
+    """CustomContainer class for being able to work
+    with the rule dictionairies easier.
+    A container is a recognised by this form :
+    {"type": "and" | "or", "components": [...]}"""
+
+    def __init__(self, operator: str, components: list, parent=None) -> None:
+        self.operator = operator
+        self.components = components
+        self.parent:'CustomContainer' = parent
+
+        if self.operator not in ["and", "or"]:
+            raise ValueError(f"Invalid operator[op={self.operator}]")
+
+    def as_dict(self):
+        """
+        Returns the dictionairy form of this object. Usefull for re-converting to json later on.
+        """
+        return {
+            'type': self.operator,
+            'components': [comp.as_dict() if isinstance(comp, CustomContainer) else comp for comp in self.components],
+        }
+
+    def __repr__(self):
+        """Used for printing the resulting object, so you dont
+        see < __main__ CustomContainer object at x567df78fd>"""
+        return str(self.as_dict())
+
+def list_equality_no_order(arr1, arr2):
+    """In python [1, 2, 3] == [3, 2, 1] equals False.
+    In our case, we just want to check if two containers
+    have the same components. This function detects any
+    elements that do not occur in both. If none are found,
+    they are identical."""
+    base = arr1
+    for k in arr2:
+        try:
+            base.remove(k)
+        except ValueError:
+            # Can't remove because it isn't there
+            return False
+    # Return not (is there anything left in base). True if empty else False
+    return not base
```

Comment:
You work around this by doing `.copy()` on the input argument in the one place you call this function, but an equality comparison that destroys one of its operands is a Bad Surprise.

(The logic here is also unnecessarily complicated -- you could just do `sorted(arr1) == sorted(arr2)`.)

Reply:
Unfortunately, sorting the list doesn't work, since it contains (at least can contain) CustomContainer classes, that can't be sorted. Hence my code. .copy() works though.